### PR TITLE
[language] implement native destroy_empty for vector

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This implements `Vector.destroy_empty` that destroys an empty vector. Will abort if the given vector is non-empty.

## Test Plan
```
cargo test
```

### CODE REVIEW

File: language/functional_tests/tests/testsuite/natives/vector/vector_resource_not_destroyed_at_return.mvir [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    resource R {}
+
+    f() {
+        let v: Vector.T<Self.R>;
+        v = Vector.empty<Self.R>();
+        return;
+    }
+}
+
+// check: RET_UNSAFE_TO_DESTROY_ERROR
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/natives/vector/vector_resource_not_destroyed_at_return.mvir]: Might be good to have a variant of this test where we let a `Vector.T<u64>` go out of scope without calling `destroy_empty`?

File: language/functional_tests/tests/testsuite/natives/vector/vector_resource_copy.mvir [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    f() {
+        let v: Vector.T<Self.R>;
+        v = Vector.empty<Self.R>();
+        Vector.destroy_empty<Self.R>(copy(v));
+        Vector.destroy_empty<Self.R>(move(v));
+        return;
+    }
+}
+
+// check: COPYLOC_RESOURCE_ERROR
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/natives/vector/vector_resource_copy.mvir]: Similar comment here--seems like this is testing that you can't copy a resource, which isn't related to the implementation of `destroy_empty`.

File: language/functional_tests/tests/testsuite/natives/vector/vector_move_after_destroy.mvir [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+main() {
+    let v: Vector.T<u64>;
+    v = Vector.empty<u64>();
+    Vector.destroy_empty<u64>(move(v));
+    v = move(v);
+    return;
+}
+
+// check: MOVELOC_UNAVAILABLE_ERROR
```
Comment (sblackshear) [inline language/functional_tests/tests/testsuite/natives/vector/vector_move_after_destroy.mvir]: Is this test related to the implementation of `destroy_empty`? It seems like this is testing that you can't double-move a resource. I can't think of a bad implementation of `destroy_empty` that would make this fail. If that's the case, maybe we should get rid of this test.

File: language/stdlib/modules/vector.mvir [hunk 4]
```diff
@@ -22,8 +22,11 @@ module Vector {
...
   native public pop_back<Element>(v: &mut Self.T<Element>): Element;
 
-  // Return true if the vector has no elements
+  // Return true if the vector has no elements.
   public is_empty<Element>(v: &Self.T<Element>): bool {
     return Self.length<Element>(move(v)) == 0;
   }
+
+  // Destroy the vector, abort if not empty.
+  native public destroy_empty<Element>(v: Self.T<Element>);
```
Comment (sblackshear) [inline language/stdlib/modules/vector.mvir]: Nit: move above `is_empty`? I think this file is slightly easier to read if we write `native` methods first and then methods with bodies?.

File: language/vm/vm_runtime/vm_runtime_types/src/value.rs [hunk 5]
```diff
@@ -50,7 +50,7 @@ mod vm_types;
 /// represented. In that sense instances of `[Value]` are what gets pushed/popped on the
 /// stack and effectively the values representable in Move.
 #[derive(PartialEq, Eq, Debug, Clone)]
-enum ValueImpl {
+pub(crate) enum ValueImpl {
```
Comment (sblackshear) [inline language/vm/vm_runtime/vm_runtime_types/src/value.rs]: This change and the one below seem unfortunate. The other native `Vector` procedure implementations didn't need to change the visibility--perhaps we can mimic their approach?

Comment (vgao1996) [reply language/vm/vm_runtime/vm_runtime_types/src/value.rs]: Yeah this is something I wasn't very sure. Planning to have a conversation with @dariorussi on this.

Comment (vgao1996) [reply language/vm/vm_runtime/vm_runtime_types/src/value.rs]: Without `pub(crate)` the function above will need to be written like this:
```
    pub fn native_destroy_empty(mut args: VecDeque<Value>) -> NativeReturnStatus {
        if let Some(v) = args.pop_front() {
            if let Some(NativeStructValue::Vector(NativeVector(v))) =
                v.value_as::<NativeStructValue>()
            {
                return if v.is_empty() {
                    NativeReturnStatus::Success {
                        cost: DESTROY_EMPTY_VEC_COST,
                        return_values: vec![],
                    }
                } else {
                    NativeReturnStatus::Aborted {
                        cost: DESTROY_EMPTY_VEC_COST,
                        error_code: DESTROY_NON_EMPTY_VEC,
                    }
                };
            }
        }
        NativeReturnStatus::InvalidArguments
    }
```
Personally I found it cumbersome compared to the nested pattern matching approach, but understand we may have a good reason to make `value_impl` private.

Comment (sblackshear) [reply language/vm/vm_runtime/vm_runtime_types/src/value.rs]: I agree that this is not as nice, but I think it's preferable to /safer than to opening up the internals of `ValueImpl`. It also ensures that the VM can change its internal representation of `ValueImpl` without breaking clients.

Comment (vgao1996) [reply language/vm/vm_runtime/vm_runtime_types/src/value.rs]: Cool. I'll get rid the `pub(crate)` for now.

Comment (dariorussi) [reply language/vm/vm_runtime/vm_runtime_types/src/value.rs]: yeah I think the way we wrote it above is more in line with how values are intended to be used and I don't think it'll be a big deal.
I wonder however if NativeStruct should have a more direct (inlined) representation in the Value story. I have not thought much about it and neither looked at the implications but it's kind of unfortunate that we have to enter a different "namespace" when it comes to native structures.
Please think about it a bit, though let's not have that block progress on this diff

**Pull request closed without merging**